From e80173eabd48d29daaf25658179d0a3e9da9d015 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 6 Feb 2025 18:51:03 +0100
Subject: [PATCH 1547/1695] drivers: stm32_iwdg: reload for stopped watchdog

For stopped watchdog, use the early interruption to reload the watchdog
without panic, the driver uses ULONG_MAX as an infinite reload indication.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git b717d1fd2d21a951bd9e9fdb1a3ce75f68d155f2 drivers: stm32_iwdg: reload for stopped watchdog]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I9f92f9b9df7227215dda8e6c1f40b4c4b877ea2a
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/435307
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Clement LE GOFFIC <clement.legoffic@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_iwdg.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/core/drivers/stm32_iwdg.c b/core/drivers/stm32_iwdg.c
index ebed7f560..4aafddb1d 100644
--- a/core/drivers/stm32_iwdg.c
+++ b/core/drivers/stm32_iwdg.c
@@ -206,11 +206,13 @@ static enum itr_return stm32_iwdg_it_handler(struct itr_handler *h)
 	 */
 	io_write32(iwdg_base + IWDG_KR_OFFSET, IWDG_KR_WPROT_KEY);
 
-	/* Disable early interrupt */
+	/* Ack early interrupt */
 	stm32_iwdg_it_ack(iwdg);
 
 	if (iwdg->nb_int > 1) {
-		iwdg->nb_int--;
+		/* Decrease interrupt counter when watchdog is not stopped*/
+		if (iwdg->nb_int < ULONG_MAX)
+			iwdg->nb_int--;
 		io_write32(get_base(iwdg) + IWDG_KR_OFFSET, IWDG_KR_RELOAD_KEY);
 	} else {
 		panic("Watchdog");
@@ -313,6 +315,10 @@ static void iwdg_wdt_stop(struct wdt_chip *chip)
 		if (iwdg->itr_handler)
 			interrupt_disable(iwdg->itr_chip, iwdg->itr_num);
 	}
+
+	/* Reload on early interrupt and no more panic */
+	iwdg->saved_nb_int = ULONG_MAX;
+	iwdg->nb_int = ULONG_MAX;
 }
 
 static void iwdg_wdt_refresh(struct wdt_chip *chip)
-- 
2.47.2

